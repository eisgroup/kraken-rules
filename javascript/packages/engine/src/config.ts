/*
 *  Copyright 2020 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { debug } from './debugger/Debugger'

/**
 * Configures Kraken evaluation. It can be configured from the console.
 * Open the console and  modify the config options
 * @example
 * global.Kraken.debug = false;
 * global.Kraken.info = false;
 * global.Kraken.debug = false;
 * @see Logger
 * @since 1.1.1
 */
export interface KrakenConfig {
    logger: {
        enabled: boolean
        debug: boolean
        info: boolean
        error: boolean
        warning: boolean
        entriesCount: number
        logs: string[]
        getLogs(): void
        clear(): void
    }
    debugger: debug.api.Debugger
}
export type WithKraken = {
    Kraken: KrakenConfig
}

function init(): void {
    const logger: KrakenConfig['logger'] = {
        enabled: process.env.NODE_ENV !== 'production',
        debug: false,
        info: true,
        error: true,
        warning: true,
        entriesCount: 2000,
        logs: [],
        clear(): void {
            const config = (global as unknown as WithKraken).Kraken
            if (config.logger.logs) {
                config.logger.logs = []
            }
        },
        getLogs(): void {
            const config = (global as unknown as WithKraken).Kraken
            // applicable only for the browser
            if (config.logger.logs && document && window) {
                const blob = new Blob([config.logger.logs.join('\n')], { type: 'text/plain' })
                const url = window.URL.createObjectURL(blob)
                const a = document.createElement('a')
                a.style.display = 'none'
                a.href = url
                a.download = `kraken-logs-${new Date().getTime()}.txt`
                document.body.appendChild(a)
                a.click()
                window.URL.revokeObjectURL(url)
                document.body.removeChild(a)
            }
        },
    }
    if (global) {
        ;(global as unknown as WithKraken).Kraken = {
            logger,
            debugger: new debug.impl.DevToolsDebugger(),
        }
        return
    }
}

function getConfig(): KrakenConfig {
    if (global) {
        return (global as unknown as { Kraken: KrakenConfig })['Kraken']
    }
    throw new Error('Kraken configuration was not initialized')
}

export const krakenConfig = { init, getConfig }
