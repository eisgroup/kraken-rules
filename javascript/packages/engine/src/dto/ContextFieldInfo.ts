/*
 *  Copyright 2020 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Contexts } from "kraken-model";
import { DataContext } from "../engine/contexts/data/DataContext";
import { ErrorCode, KrakenRuntimeError } from "../error/KrakenRuntimeError";

export class ContextFieldInfo {
    readonly contextId: string;
    readonly contextName: string;
    readonly fieldName: string;
    readonly fieldPath: string;
    readonly fieldType: string;
    constructor(dataContext: DataContext, fieldName: string) {
        let projection: Contexts.ContextField;
        if (dataContext.definitionProjection && dataContext.definitionProjection[fieldName]) {
            projection = dataContext.definitionProjection[fieldName];
        } else {
            throw new KrakenRuntimeError(
                ErrorCode.INCORRECT_MODEL_TREE,
                `Fields projection for a DataContext instance lacks of field '${fieldName}'`
            );
        }
        this.contextId = dataContext.contextId;
        this.contextName = dataContext.contextName;
        this.fieldName = fieldName;
        this.fieldPath = projection.fieldPath;
        this.fieldType = projection.fieldType;
    }
}
