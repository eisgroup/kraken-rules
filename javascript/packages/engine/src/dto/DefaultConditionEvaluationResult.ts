/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { ConditionEvaluation, ConditionEvaluationResult, ExpressionEvaluationResult } from 'kraken-engine-api'

export class DefaultConditionEvaluationResult implements ConditionEvaluationResult {
    private constructor(
        public readonly conditionEvaluation: ConditionEvaluation,
        public readonly error: ExpressionEvaluationResult.ErrorResult | undefined,
    ) {}

    static fromError = (error: ExpressionEvaluationResult.ErrorResult) => {
        return new DefaultConditionEvaluationResult(ConditionEvaluation.ERROR, error)
    }

    static of = (evaluation: ConditionEvaluation) => new DefaultConditionEvaluationResult(evaluation, undefined)
}

export const conditionEvaluationTypeChecker = {
    isApplicable(result: ConditionEvaluationResult): boolean {
        return result.conditionEvaluation === ConditionEvaluation.APPLICABLE
    },
}
