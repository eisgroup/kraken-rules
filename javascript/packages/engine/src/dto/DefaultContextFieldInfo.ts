/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Contexts } from 'kraken-model'
import { DataContext } from '../engine/contexts/data/DataContext'
import { CONTEXT_MODEL_TREE_MISSING_FIELD, KrakenRuntimeError, SystemMessageBuilder } from '../error/KrakenRuntimeError'
import { ContextFieldInfo } from 'kraken-engine-api'

export class DefaultContextFieldInfo implements ContextFieldInfo {
    readonly contextId: string
    readonly contextName: string
    readonly fieldName: string
    readonly fieldPath: string
    readonly fieldType: string

    constructor(dataContext: DataContext, fieldName: string) {
        let contextField: Contexts.ContextField
        if (dataContext.contextDefinition.fields?.[fieldName]) {
            contextField = dataContext.contextDefinition.fields[fieldName]
        } else {
            const m = new SystemMessageBuilder(CONTEXT_MODEL_TREE_MISSING_FIELD).parameters(fieldName).build()
            throw new KrakenRuntimeError(m)
        }
        this.contextId = dataContext.contextId
        this.contextName = dataContext.contextName
        this.fieldName = fieldName
        this.fieldPath = contextField.fieldPath
        this.fieldType = contextField.fieldType
    }
}
