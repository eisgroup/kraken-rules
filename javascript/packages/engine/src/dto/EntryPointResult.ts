/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from "./FieldEvaluationResult";
import { RuleEvaluationResults } from "./RuleEvaluationResults";
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult;

import { Reducer } from "declarative-js";
import flat = Reducer.flat;
import { ConditionEvaluationResult } from "./ConditionEvaluationResult";

/**
 * Contains evaluation results from specific entry point
 */
export class EntryPointResult {
    private readonly results: Record<string, FieldEvaluationResult>;
    readonly evaluationTimestamp: Date;
    constructor(results: Record<string, FieldEvaluationResult>) {
        this.results = results;
        this.getApplicableResults.bind(this);
        this.evaluationTimestamp = new Date();
    }
    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     */
    getAllRuleResults(): RuleEvaluationResult[] {
        return Object.values(this.results)
            .map(x => x.ruleResults)
            .reduce(flat, []);
    }
    getFieldResults(): Record<string, FieldEvaluationResult> {
        return this.results;
    }

    getApplicableResults(): RuleEvaluationResult[] {
        return this.getAllRuleResults()
            .filter(r => ConditionEvaluationResult.isApplicable(r.conditionEvaluationResult));
    }

    static empty(): EntryPointResult {
        return new EntryPointResult({});
    }
}
