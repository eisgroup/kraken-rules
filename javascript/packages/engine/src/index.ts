/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { krakenConfig } from './config'
krakenConfig.init()

// spi
export * from './engine/contexts/info/DataObjectInfoResolver'
export * from './engine/contexts/info/ContextInstanceInfoResolver'

// engine
export * from './engine/executer/SyncEngine'

export * from './bundle-cache/EntryPointBundleCache'
export * from './bundle-cache/delta-cache/RepoClientCache'
export * from './bundle-cache/delta-cache/DeltaBundleCache'
export * from './bundle-cache/dimension-set-cache/DimensionSetBundleCache'
export * from './bundle-cache/dimension-set-cache/DimensionSetCacheLogger'
export * from './bundle-cache/expression-context-manager/ExpressionContextManager'
export * from './bundle-cache/expression-context-manager/ExpressionContextManagerImpl'

export { registry, KelFunction, FunctionParameter } from './engine/runtime/expressions/ExpressionEvaluator'
export { FunctionScope } from './engine/runtime/expressions/functionLibrary/Registry'
export * from './engine/runtime/EvaluationMode'

// Payload result utils
export * from './engine/results/PayloadResultCreator'
export * from './engine/results/PayloadResultTypeChecker'
export { ValueChangedEvent } from './engine/results/ValueChangedEvent'

// Condition result utils
export { conditionEvaluationTypeChecker } from './dto/DefaultConditionEvaluationResult'

export { DataContext } from './engine/contexts/data/DataContext'

// field metadata reducer
export * from './engine/results/field_metadata_reducer/FieldErrorMessage'

export * from './models/EntryPointBundle'
export * from './models/ContextModelTree'

// reducer
export * from './engine/results/RuleOverrideContextExtractor'
export * from './engine/results/Reducer'
export * from './engine/results/Localization'

// field metadata reducer
export * from './engine/results/field_metadata_reducer/FieldMetadataReducer'

// error
export * from './error/KrakenRuntimeError'

// re-export
export {
    ContextInstanceInfo,
    RuleOverride,
    RuleInfo,
    ContextFieldInfo,
    EntryPointResult,
    ConditionEvaluationResult,
    FieldEvaluationResult,
    RuleEvaluationResults,
    PayloadResultType,
    PayloadResult,
    ErrorAwarePayloadResult,
    ErrorMessage,
    ValidationPayloadResult,
    AccessibilityPayloadResult,
    VisibilityPayloadResult,
    DefaultValuePayloadResult,
    AssertionPayloadResult,
    SizePayloadResult,
    SizeRangePayloadResult,
    LengthPayloadResult,
    RegExpPayloadResult,
    UsagePayloadResult,
    ValueListPayloadResult,
    RuleEvent,
    FieldMetadata,
    FieldMetadataResult,
    ConditionEvaluation,
} from 'kraken-engine-api'
