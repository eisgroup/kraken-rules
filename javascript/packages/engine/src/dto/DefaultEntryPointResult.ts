/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Reducer } from 'declarative-js'
import flat = Reducer.flat

import { EntryPointResult, FieldEvaluationResult, RuleEvaluationResults } from 'kraken-engine-api'

import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult
import { conditionEvaluationTypeChecker } from './DefaultConditionEvaluationResult'

export class DefaultEntryPointResult implements EntryPointResult {
    readonly results: Record<string, FieldEvaluationResult>
    readonly evaluationTimestamp: Date
    readonly ruleTimezoneId: string

    constructor(results: Record<string, FieldEvaluationResult>, evaluationTimestamp: Date, ruleTimezoneId: string) {
        this.results = results
        this.getApplicableResults.bind(this)
        this.evaluationTimestamp = evaluationTimestamp
        this.ruleTimezoneId = ruleTimezoneId
    }

    getAllRuleResults(): RuleEvaluationResult[] {
        return Object.values(this.results)
            .map(x => x.ruleResults)
            .reduce(flat, [])
    }

    getFieldResults(): Record<string, FieldEvaluationResult> {
        return this.results
    }

    getApplicableResults(): RuleEvaluationResult[] {
        return this.getAllRuleResults().filter(r =>
            conditionEvaluationTypeChecker.isApplicable(r.conditionEvaluationResult),
        )
    }
}
