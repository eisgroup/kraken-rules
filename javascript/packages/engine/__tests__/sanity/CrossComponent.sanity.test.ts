/* eslint-disable @typescript-eslint/no-non-null-assertion */
/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { sanityMocks } from './_AutoPolicyObject.mocks'
import { RuleEvaluationResults, ValidationPayloadResult } from 'kraken-engine-api'
import { sanityEngine, sanityEngineExtendedPolicy } from './_SanityEngine'
import { EntryPointName } from './_SanityEntryPointNames'
import { conditionEvaluationTypeChecker } from '../../src/dto/DefaultConditionEvaluationResult'
import { MatchableResults } from 'kraken-jest-matchers'
import { FieldMetadataReducer } from '../../src'

describe('Cross context references sanity tests', () => {
    const { valid, empty, model, emptyExtended } = sanityMocks
    const expectEvaluationToBe = ({
        model: policy,
        entryPoint,
        payloadSuccess,
        restriction,
    }: {
        model: object
        entryPoint: EntryPointName
        payloadSuccess: boolean
        restriction?: object
    }) => {
        const results = restriction
            ? sanityEngine.evaluateSubTree(policy, restriction, entryPoint)
            : sanityEngine.evaluate(policy, entryPoint)
        expect(results).not.k_toHaveExpressionsFailures()
        expect(results.getAllRuleResults()).toHaveLength(1)
        expect(
            (
                (results.getAllRuleResults()[0] as RuleEvaluationResults.ApplicableRuleEvaluationResult)
                    .payloadResult as ValidationPayloadResult
            ).success,
        ).toBe(payloadSuccess)
    }
    it('should find reference from 2nd level to root (condition false)', () => {
        const data = valid()
        data['policyCurrency'] = 'EUR'
        const results = sanityEngine.evaluate(data, 'Cross-1')

        expect(results.getAllRuleResults()).toHaveLength(1)
        const conditionEvaluationResult = results.getAllRuleResults()[0].conditionEvaluationResult
        const isRuleApplicable = conditionEvaluationTypeChecker.isApplicable(conditionEvaluationResult)
        expect(isRuleApplicable).toBeFalsy()
        expect(results).not.k_toHaveExpressionsFailures()
    })
    it('should find reference from 2nd level to root (condition true)', () => {
        const data = valid()
        data.policyCurrency = 'USD'
        const results = sanityEngine.evaluate(data, 'Cross-1')

        expect(results.getAllRuleResults()).toHaveLength(1)
        const conditionEvaluationResult = results.getAllRuleResults()[0].conditionEvaluationResult
        const isRuleApplicable = conditionEvaluationTypeChecker.isApplicable(conditionEvaluationResult)
        expect(isRuleApplicable).toBeTruthy()
        expect(results).not.k_toHaveExpressionsFailures()
    })
    it('should find reference from 2nd level to root (condition true) - ExtendedModel', () => {
        const data = empty()
        data['policyCurrency'] = 'USD'
        const results = sanityEngine.evaluate(data, 'Cross-1')

        expect(results).not.k_toHaveExpressionsFailures()
        expect(results.getAllRuleResults()).toHaveLength(1)
        const conditionEvaluationResult = results.getAllRuleResults()[0].conditionEvaluationResult
        const isRuleApplicable = conditionEvaluationTypeChecker.isApplicable(conditionEvaluationResult)
        expect(isRuleApplicable).toBeTruthy()
    })
    it('should set default value from Policy to CreditCard', () => {
        const data = empty()
        data['policyCurrency'] = '1'
        const results = sanityEngine.evaluate(data, 'R-CCR-default-CreditCardInfo-fromAutoPolicy')

        expect(results).not.k_toHaveExpressionsFailures()
        expect(results.getAllRuleResults()).toHaveLength(1)
        expect(data.billingInfo!.creditCardInfo!.cardNumber).toBe('1')
    })
    it('should reference from 3rd level to root', () => {
        expectEvaluationToBe({
            model: model({ policyCurrency: 'a', bilAdr_countryCd: 'a' }),
            entryPoint: 'R-CCR-assert-BillingAddress',
            payloadSuccess: true,
        })
        expectEvaluationToBe({
            model: model({ policyCurrency: 'a', bilAdr_countryCd: 'b' }),
            entryPoint: 'R-CCR-assert-BillingAddress',
            payloadSuccess: false,
        })
    })
    it('should reference from 3rd level to 2nd level', () => {
        expectEvaluationToBe({
            model: model({ creditCardInfo_cardType: 'a', bilAdr_countryCd: 'a' }),
            entryPoint: 'R-CCR-assert-BillingAddress-toCreditCard',
            payloadSuccess: true,
        })
        expectEvaluationToBe({
            model: model({ creditCardInfo_cardType: 'a', bilAdr_countryCd: 'b' }),
            entryPoint: 'R-CCR-assert-BillingAddress-toCreditCard',
            payloadSuccess: false,
        })
    })
    it('should reference from 2nd level collection to root', () => {
        expectEvaluationToBe({
            model: model({ policyCurrency: 'a', vehModel: 'a' }),
            entryPoint: 'R-CCR-assert-Vehicle-toAutoPolicy',
            payloadSuccess: true,
        })
        expectEvaluationToBe({
            model: model({ creditCardInfo_cardType: 'a', policyCurrency: 'b' }),
            entryPoint: 'R-CCR-assert-Vehicle-toAutoPolicy',
            payloadSuccess: false,
        })
    })
    it('should reference from root to 2nd level', () => {
        expectEvaluationToBe({
            model: model({ creditCardInfo_cardType: 'a', policyCurrency: 'a' }),
            entryPoint: 'R-CCR-assert-AutoPolicy-toCreditCard',
            payloadSuccess: true,
        })
        expectEvaluationToBe({
            model: model({ creditCardInfo_cardType: 'a', policyCurrency: 'b' }),
            entryPoint: 'R-CCR-assert-AutoPolicy-toCreditCard',
            payloadSuccess: false,
        })
    })
    it('should reference from 3rd to 3rd level same branch', () => {
        expectEvaluationToBe({
            model: model({ personInfo_name: 'a', partyDriverType: 'a' }),
            entryPoint: 'R-CCR-assert-DriverInfo-PersonInfo',
            payloadSuccess: true,
        })
        expectEvaluationToBe({
            model: model({ personInfo_name: 'a', partyDriverType: 'b' }),
            entryPoint: 'R-CCR-assert-DriverInfo-PersonInfo',
            payloadSuccess: false,
        })
    })
    it('should reference from 3rd to 2nd level different branch', () => {
        expectEvaluationToBe({
            model: model({ creditCardInfo_cardType: 'a', partyDriverType: 'a' }),
            entryPoint: 'R-CCR-assert-DriverInfo-CreditCardInfo',
            payloadSuccess: true,
        })
        expectEvaluationToBe({
            model: model({ creditCardInfo_cardType: 'b', partyDriverType: 'a' }),
            entryPoint: 'R-CCR-assert-DriverInfo-CreditCardInfo',
            payloadSuccess: false,
        })
    })
    it('should reference from 3rd to 2nd level different branch with a restriction - ExtendedModel', () => {
        expectEvaluationToBe({
            model: model({ partyDriverType: 'a', creditCardInfo_cardType: 'a' }),
            entryPoint: 'R-CCR-assert-DriverInfo-CreditCardInfo',
            payloadSuccess: true,
        })
        expectEvaluationToBe({
            model: model({ partyDriverType: 'a', creditCardInfo_cardType: 'b' }),
            entryPoint: 'R-CCR-assert-DriverInfo-CreditCardInfo',
            payloadSuccess: false,
        })
    })
    it('should reference from 3rd to 2nd level different branch with restriction', () => {
        const data = model({ partyDriverType: 'a', creditCardInfo_cardType: 'a' })
        expectEvaluationToBe({
            model: data,
            entryPoint: 'R-CCR-assert-DriverInfo-CreditCardInfo',
            payloadSuccess: true,
            restriction: data.parties![0],
        })
        const data2 = model({ partyDriverType: 'a', creditCardInfo_cardType: 'b' })
        expectEvaluationToBe({
            model: data2,
            entryPoint: 'R-CCR-assert-DriverInfo-CreditCardInfo',
            payloadSuccess: false,
            restriction: data2.parties![0],
        })
    })
    it('should access reference with multiple cardinality', () => {
        const policy = empty()
        policy.parties = [
            {
                cd: 'Party',
                id: 'p01',
                roles: [
                    { cd: 'PartyRole', id: 'c01', role: 'a' },
                    { cd: 'PartyRole', id: 'c02', role: 'b' },
                ],
            },
            {
                cd: 'Party',
                id: 'p02',
                roles: [
                    { cd: 'PartyRole', id: 'c03', role: 'c' },
                    { cd: 'PartyRole', id: 'c04', role: 'd' },
                ],
            },
        ]
        const results = sanityEngine.evaluate(policy, 'R-CCR-Policy-PartyRole')
        const fieldResults = new FieldMetadataReducer().reduce(results)
        const matchableResults = {
            entryPointResults: results,
            fieldResults,
        } as MatchableResults

        expect(results).k_toMatchResultsStats({ total: 1, critical: 0 })
        expect(results).k_toHaveExpressionsFailures(0)
        expect(matchableResults).k_toMatchResultsSnapshots()
    })
    it('should check is self reference with inherited context will be skipped', () => {
        const policy = empty()
        policy.policyNumber = 'P01'
        const results = sanityEngine.evaluate(policy, 'R-CCR-Policy-ExPolicy')
        const fieldResults = new FieldMetadataReducer().reduce(results)
        const matchableResults = {
            entryPointResults: results,
            fieldResults,
        } as MatchableResults

        expect(results).k_toMatchResultsStats({ total: 1, critical: 0 })
        expect(results).k_toHaveExpressionsFailures(0)
        expect(matchableResults).k_toMatchResultsSnapshots()
    })
    it('should not fail on executing rules on same instance with inheritance rules', () => {
        const policy = emptyExtended()
        policy.policyCurrency = 'USD'
        policy.billingInfo!.creditCardInfo!.cardType = 'USD'
        const results = sanityEngineExtendedPolicy.evaluate(policy, 'Inheritance-CCR')
        const fieldResults = new FieldMetadataReducer().reduce(results)
        const matchableResults = {
            entryPointResults: results,
            fieldResults,
        } as MatchableResults

        expect(results).k_toMatchResultsStats({ total: 2, critical: 0 })
        expect(results).k_toHaveExpressionsFailures(0)
        expect(matchableResults).k_toMatchResultsSnapshots()
    })
})
