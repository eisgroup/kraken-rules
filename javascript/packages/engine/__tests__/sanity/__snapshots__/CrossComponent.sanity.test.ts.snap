// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cross context references sanity tests should access reference with multiple cardinality 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:policyNumber": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "policyNumber",
          "fieldPath": "policyNumber",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": Object {
                "errorCode": "code",
                "errorMessage": "Value didn't match assertion",
                "templateVariables": Array [],
              },
              "success": true,
              "type": 1,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "ASSERTION",
              "ruleName": "R-CCR-Policy-PartyRole",
              "targetPath": "policyNumber",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "Policy:0:policyNumber": Object {
      "fieldType": "STRING",
      "id": "Policy:0:policyNumber",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyNumber",
      "ruleResults": Array [
        Object {
          "errorCode": "code",
          "errorMessage": "Value didn't match assertion",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": Object {
              "errorCode": "code",
              "errorMessage": "Value didn't match assertion",
              "templateVariables": Array [],
            },
            "success": true,
            "type": 1,
            "validationSeverity": "critical",
          },
          "ruleName": "R-CCR-Policy-PartyRole",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
  },
}
`;

exports[`Cross context references sanity tests should check is self reference with inherited context will be skipped 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:0:policyCurrency": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "policyCurrency",
          "fieldPath": "policyCurrency",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": Object {
                "errorCode": "code",
                "errorMessage": "Value didn't match assertion",
                "templateVariables": Array [],
              },
              "success": true,
              "type": 1,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "ASSERTION",
              "ruleName": "R-CCR-Policy-ExPolicy",
              "targetPath": "policyCurrency",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "Policy:0:policyCurrency": Object {
      "fieldType": "STRING",
      "id": "Policy:0:policyCurrency",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyCurrency",
      "ruleResults": Array [
        Object {
          "errorCode": "code",
          "errorMessage": "Value didn't match assertion",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": Object {
              "errorCode": "code",
              "errorMessage": "Value didn't match assertion",
              "templateVariables": Array [],
            },
            "success": true,
            "type": 1,
            "validationSeverity": "critical",
          },
          "ruleName": "R-CCR-Policy-ExPolicy",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
  },
}
`;

exports[`Cross context references sanity tests should not fail on executing rules on same instance with inheritance rules 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "PolicyExtended:0:policyCurrency": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "PolicyExtended",
          "fieldName": "policyCurrency",
          "fieldPath": "policyCurrency",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "success": true,
              "type": 1,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "PolicyExtended",
              "payloadtype": "ASSERTION",
              "ruleName": "Inheritance-CCR-02",
              "targetPath": "policyCurrency",
            },
          },
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "success": true,
              "type": 1,
              "validationSeverity": "critical",
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "ASSERTION",
              "ruleName": "Inheritance-CCR-01",
              "targetPath": "policyCurrency",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "PolicyExtended:0:policyCurrency": Object {
      "fieldType": "STRING",
      "id": "PolicyExtended:0:policyCurrency",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyCurrency",
      "ruleResults": Array [
        Object {
          "errorCode": "rule-assertion-error",
          "errorMessage": "Assertion failed",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "success": true,
            "type": 1,
            "validationSeverity": "critical",
          },
          "ruleName": "Inheritance-CCR-02",
          "severity": "critical",
          "templateVariables": Array [],
        },
        Object {
          "errorCode": "rule-assertion-error",
          "errorMessage": "Assertion failed",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "success": true,
            "type": 1,
            "validationSeverity": "critical",
          },
          "ruleName": "Inheritance-CCR-01",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
  },
}
`;
